qtfred/src/ui/dialogs/ShipEditor/ShipCustomWarpDialog.cpp

```diff
@@ -0,0 +1,212 @@
+	: QDialog(parent), ui(new Ui::ShipCustomWarpDialog()),
+	  _model(new ShipCustomWarpDialogModel(this, viewport, departure)), _viewport(viewport)
+{
+	ui->setupUi(this);
+	connect(this, &QDialog::accepted, _model.get(), &ShipCustomWarpDialogModel::apply);
+
+	connect(ui->buttonBox, &QDialogButtonBox::rejected, this, &ShipCustomWarpDialog::rejectHandler);
+
+	connect(_model.get(), &AbstractDialogModel::modelChanged, this, [this]() { updateUI(false); });
+
+	connect(ui->comboBoxType,
+		static_cast<void (QComboBox::*)(int)>(&QComboBox::currentIndexChanged),
+		_model.get(),
+		&ShipCustomWarpDialogModel::setType);
+	connect(ui->lineEditStartSound, &QLineEdit::editingFinished, this, &ShipCustomWarpDialog::startSoundChanged);
+	connect(ui->lineEditEndSound, &QLineEdit::editingFinished, this, &ShipCustomWarpDialog::endSoundChanged);
+	connect(ui->doubleSpinBoxEngage,
+		QOverload<double>::of(&QDoubleSpinBox::valueChanged),
+		_model.get(),
+		&ShipCustomWarpDialogModel::setEngageTime);
+	connect(ui->doubleSpinBoxSpeed,
+		QOverload<double>::of(&QDoubleSpinBox::valueChanged),
+		_model.get(),
+		&ShipCustomWarpDialogModel::setSpeed);
+	connect(ui->doubleSpinBoxTime,
+		QOverload<double>::of(&QDoubleSpinBox::valueChanged),
+		_model.get(),
+		&ShipCustomWarpDialogModel::setTime);
+	connect(ui->doubleSpinBoxExponent,
+		QOverload<double>::of(&QDoubleSpinBox::valueChanged),
+		_model.get(),
+		&ShipCustomWarpDialogModel::setExponent);
+	connect(ui->doubleSpinBoxRadius,
+		QOverload<double>::of(&QDoubleSpinBox::valueChanged),
+		_model.get(),
+		&ShipCustomWarpDialogModel::setRadius);
+	connect(ui->checkBoxSupercap, &QCheckBox::toggled, _model.get(), [=](bool param) {
+		static_cast<ShipCustomWarpDialogModel*>(_model.get())->setSupercap(param);
+	});
+	connect(ui->lineEditAnim, &QLineEdit::editingFinished, this, &ShipCustomWarpDialog::animChanged);
+	connect(ui->doubleSpinBoxPlayerSpeed,
+		QOverload<double>::of(&QDoubleSpinBox::valueChanged),
+		_model.get(),
+		&ShipCustomWarpDialogModel::setPlayerSpeed);
+
+	if (departure) {
+		this->setWindowTitle("Edit Warp-Out Parameters");
+	} else {
+		this->setWindowTitle("Edit Warp-in Parameters");
+	}
+	updateUI(true);
+	// Resize the dialog to the minimum size
+	resize(QDialog::sizeHint());
+}
+
+ShipCustomWarpDialog::~ShipCustomWarpDialog() = default;
+void ShipCustomWarpDialog::closeEvent(QCloseEvent* e)
+{
+	if (_model->query_modified()) {
+		auto button = _viewport->dialogProvider->showButtonDialog(DialogType::Question,
+			"Changes detected",
+			"Do you want to keep your changes?",
+			{DialogButton::Yes, DialogButton::No, DialogButton::Cancel});
+
+		if (button == DialogButton::Cancel) {
+			e->ignore();
+			return;
+		}
+
+		if (button == DialogButton::Yes) {
+			accept();
+			return;
+		}
+		if (button == DialogButton::No) {
+			_model->reject();
+		}
+	}
+	QDialog::closeEvent(e);
+}
+void ShipCustomWarpDialog::updateUI(const bool firstrun)
```

Comment:
In this function, rather than using string comparisons to the warp type name, it would be better to compare the warp type index (e.g. `i` or `ui->comboBoxType->currentIndex()`) to one of the `WT_` definitions in shipfx.h.  This makes the code more maintainable.  (It's also slightly faster.)